test/provider/model/ClientRegistry.unittest.ts

```diff
@@ -0,0 +1,153 @@
+});
+
+describe('When attemping to launch an app through the client registry', () => {
+    const storedApp: StoredApplication = {
+        type: 'manifest',
+        id: 'cr-test-app',
+        manifestUrl: 'some-manifest-url'
+    };
+
+    const mockApiHandler = createMockApiHandler();
+    const mockEnvironment = createMockEnvironment();
+
+    let clientRegistry: ClientRegistry;
+    let database: Database;
+
+    beforeEach(async () => {
+        const {Database: ActualDatabase} = jest.requireActual('../../../src/provider/model/database/Database');
+        database = await (new ActualDatabase() as Database).delayedInit();
+
+        const collection = database.get(CollectionMap.APPLICATIONS);
+        await collection.upsert(storedApp);
+
+        clientRegistry = new ClientRegistry(mockApiHandler, database, mockEnvironment);
+    });
+
+    afterEach(async () => {
+        const collection = database.get(CollectionMap.APPLICATIONS);
+        await collection.delete((await collection.getAll()).map(app => app.id));
+    });
+
+    test('If the app is not running, the client registry will try to start the app from the stored data in the database', async () => {
+        mockEnvironment.isApplicationRunning.mockImplementation(async () => {
+            return false;
+        });
+
+        await clientRegistry.tryLaunchApplication(storedApp.id);
+
+        expect(mockEnvironment.startApplication).toBeCalledTimes(1);
+        expect(mockEnvironment.startApplication).toBeCalledWith(storedApp);
+    });
+
+    test('If the app is already running, the client registry will not attempt to start the app', async () => {
+        mockEnvironment.isApplicationRunning.mockImplementation(async () => {
+            return true;
+        });
+
+        await clientRegistry.tryLaunchApplication(storedApp.id);
+
+        expect(mockEnvironment.startApplication).toBeCalledTimes(0);
+    });
+});
+
+describe('When querying windows', () => {
+    const mockUuid1 = 'mock-app-1';
+    const mockUuid2 = 'mock-app-2';
+
+    const mockApp1Window1 = {uuid: mockUuid1, name: 'mock-window-1'};
+    const mockApp1Window2 = {uuid: mockUuid1, name: 'mock-window-2'};
+
+    const mockApp2Window1 = {uuid: mockUuid2, name: 'mock-window-1'};
+
+    const mockApiHandler = createMockApiHandler();
+    const mockEnvironment = createMockEnvironment();
+    const mockDatabase = createMockDatabase();
+
+    let clientRegistry: ClientRegistry;
+
+    beforeEach(() => {
+        clientRegistry = new ClientRegistry(mockApiHandler, mockDatabase, mockEnvironment);
+    });
+
+    test('When querying windows for a given app, the expected connected windows are returned', () => {
+        mockApiHandler.getClientConnections.mockReturnValue([mockApp1Window1, mockApp2Window1]);
+
+        expect(clientRegistry.getAllAppWindows(mockUuid1)).toEqual([mockApp1Window1]);
+    });
+
+    test('When querying if an app is action-ready, the expected value is returned', () => {
```

Comment:
This test seems like it's actually lots of separate tests all squashed into one. Can it be separated out and use describes as needed for any shared state

Reply:
Or better yet, maybe a `test.each` could work here?

Reply:
✅